src/libData/BlockChainData/BlockChain.h

```diff
@@ -0,0 +1,160 @@
+/**
+* Copyright (c) 2018 Zilliqa 
+* This source code is being disclosed to you solely for the purpose of your participation in 
+* testing Zilliqa. You may view, compile and run the code for that purpose and pursuant to 
+* the protocols and algorithms that are programmed into, and intended by, the code. You may 
+* not do anything else with the code without express permission from Zilliqa Research Pte. Ltd., 
+* including modifying or publishing the code (or any part of it), and developing or forming 
+* another public or private blockchain network. This source code is provided ‘as is’ and no 
+* warranties are given as to title or non-infringement, merchantability or fitness for purpose 
+* and, to the extent permitted by law, all liability for your use of the code is disclaimed. 
+* Some programs in this code are governed by the GNU General Public License v3.0 (available at 
+* https://www.gnu.org/licenses/gpl-3.0.en.html) (‘GPLv3’). The programs that are governed by 
+* GPLv3.0 are those programs that are located in the folders src/depends and tests/depends 
+* and which include a reference to GPLv3 in their program files.
+**/
+
+#ifndef __BLOCKCHAIN_H__
+#define __BLOCKCHAIN_H__
+
+#include <mutex>
+#include <vector>
+
+#include <boost/multiprecision/cpp_int.hpp>
+
+#include "libData/BlockData/Block/DSBlock.h"
+#include "libData/DataStructures/CircularArray.h"
+#include "libPersistence/BlockStorage.h"
+
+/// Transient storage for DS/Tx/VC blocks.
+template<class T> class BlockChain
+{
+    std::mutex m_mutexBlocks;
+    CircularArray<T> m_blocks;
+
+protected:
+    /// Constructor.
+    BlockChain() { Reset(); }
+
+    virtual T GetBlockFromPersistentStorage(
+        const boost::multiprecision::uint256_t& blockNum)
+        = 0;
+
+public:
+    /// Destructor.
+    ~BlockChain() {}
+
+    /// Reset
+    void Reset() { m_blocks.resize(BLOCKCHAIN_SIZE); }
+
+    /// Returns the number of blocks.
+    boost::multiprecision::uint256_t GetBlockCount()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        return m_blocks.size();
+    }
+
+    /// Returns the last stored block.
+    T GetLastBlock()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        try
+        {
+            return m_blocks.back();
+        }
+        catch (...)
```

Comment:
I didn't find this try-catch in the original code but it's a good fix since `CircularArray::back()` is throwing [something](https://github.com/Zilliqa/Zilliqa/blob/c2b49315a37f36b035d1576ced796b728cc6ca26/src/libData/DataStructures/CircularArray.h#L90-L98). However, from the aspect of interface design I feel that `CircularArray::back` is doing too much (detecting empty-array error), since the caller could just call something like `.empty()` beforehand instead of going through the exception stuff.

In the standard library, for example, [`vector::back`](http://en.cppreference.com/w/cpp/container/vector/back) is not handling the empty container error and also not throwing any exception. (It actually leaves it as a UB)